File: sdk/bundled/src/main/java/ink/reactor/sdk/bundled/config/yaml/SectionToYamlConverter.java
package ink.reactor.sdk.bundled.config.yaml;

import ink.reactor.sdk.config.ConfigSection;
import lombok.RequiredArgsConstructor;

import java.util.Collection;
import java.util.Map;
import java.util.Set;

@RequiredArgsConstructor
public final class SectionToYamlConverter {

    private final int indentSpaces;
    private final StringBuilder builder = new StringBuilder();
    private int indentLevel = -1;

    public String toYaml(final ConfigSection configSection) {
        appendSection(configSection);
        return builder.toString();
    }

    private void appendSection(final ConfigSection subSection) {
        indentLevel++;

        final Set<? extends Map.Entry<String, ?>> entries = subSection.getData().entrySet();
        final int size = entries.size();
        int i = 0;

        for (final Map.Entry<String, ?> entry : entries) {
            final String key = entry.getKey();
            final Object value = entry.getValue();
            builder.repeat(' ', indentLevel * indentSpaces);
            builder.append(key).append(':').append(' ');
            appendObject(value);
            if (++i != size) {
                builder.append('\n');
            }
        }

        indentLevel--;
    }

    private void appendObject(final Object value) {
        switch (value) {
            case null -> builder.append("null");
            case Boolean bool -> builder.append(bool.booleanValue());
            case ConfigSection subSection -> {
                builder.append('\n');
                appendSection(subSection);
            }
            case Number _ -> builder.append(value);
            case Collection<?> collection -> appendCollection(collection);
            default -> builder.append('"').append(value).append('"');
        }
    }

    private void appendCollection(final Collection<?> collection) {
        if (collection.isEmpty()) {
            builder.append("[]");
            return;
        }

        builder.append('\n');
        final int size = collection.size();
        int i = 0;
        for (final Object o : collection) {
            builder.repeat(' ', indentSpaces * (indentLevel+1)).append('-').append(' ');
            appendObject(o);
            if (++i != size) {
                builder.append('\n');
            }
        }
    }
}

File: sdk/bundled/src/main/kotlin/ink/reactor/sdk/bundled/config/yaml/SectionToYamlConverter.kt
package ink.reactor.sdk.bundled.config.yaml

import ink.reactor.sdk.config.ConfigSection

class SectionToYamlConverter(private val indentSpaces: Int) {
    private val builder = StringBuilder()
    private var indentLevel = 0

    fun toYaml(section: ConfigSection): String {
        appendMap(section.data)
        return builder.toString()
    }

    private fun appendMap(map: Map<String, Any?>) {
        map.entries.forEachIndexed { index, entry ->
            val (key, value) = entry

            builder.append(" ".repeat(indentLevel * indentSpaces))
            builder.append(key).append(": ")

            appendObject(value)

            if (index < map.size - 1) builder.append("\n")
        }
    }

    private fun appendObject(value: Any?) {
        when (value) {
            null -> builder.append("null")
            is Boolean, is Number -> builder.append(value)
            is ConfigSection -> {
                builder.append("\n")
                indentLevel++
                appendMap(value.data)
                indentLevel--
            }
            is Map<*, *> -> {
                builder.append("\n")
                indentLevel++
                @Suppress("UNCHECKED_CAST")
                appendMap(value as Map<String, Any?>)
                indentLevel--
            }
            is Collection<*> -> appendCollection(value)
            else -> {
                val escaped = value.toString()
                    .replace("\\", "\\\\")
                    .replace("\"", "\\\"")
                builder.append("\"").append(escaped).append("\"")
            }
        }
    }

    private fun appendCollection(collection: Collection<*>) {
        if (collection.isEmpty()) {
            builder.append("[]")
            return
        }
        builder.append("\n")
        collection.forEachIndexed { index, item ->
            builder.append(" ".repeat(indentLevel * indentSpaces)).append("- ")
            indentLevel++
            appendObject(item)
            indentLevel--
            if (index < collection.size - 1) builder.append("\n")
        }
    }
}
